routers/api/v1/repo/pull.go

```diff
@@ -597,22 +597,41 @@ func MergePullRequest(ctx *context.APIContext, form auth.MergePullRequestForm) {
 		return
 	}
 
-	if !pr.CanAutoMerge() || pr.HasMerged || pr.IsWorkInProgress() {
-		ctx.Status(http.StatusMethodNotAllowed)
+	perm, err := models.GetUserRepoPermission(ctx.Repo.Repository, ctx.User)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, "GetUserRepoPermission", err)
 		return
 	}
 
-	isPass, err := pull_service.IsPullCommitStatusPass(pr)
+	allowedMerge, err := pull_service.IsUserAllowedToMerge(pr, perm, ctx.User)
 	if err != nil {
-		ctx.Error(http.StatusInternalServerError, "IsPullCommitStatusPass", err)
+		ctx.Error(http.StatusInternalServerError, "IsUSerAllowedToMerge", err)
+		return
+	}
+	if !allowedMerge {
+		ctx.Error(http.StatusMethodNotAllowed, "Merge", "User not allowed to merge PR")
 		return
 	}
 
-	if !isPass && !ctx.IsUserRepoAdmin() {
+	if !pr.CanAutoMerge() || pr.HasMerged || pr.IsWorkInProgress() {
 		ctx.Status(http.StatusMethodNotAllowed)
 		return
 	}
 
+	if err := pull_service.CheckPrReadyToMerge(pr); err != nil {
```

Comment:
We should check whether the error is actually a check failed problem instead of -for example- a database or git problem (e.g. `models.IsErrNotAllowedToMerge(err)`).

Reply:
fixed